citest/reporting/html_document_manager.py

```diff
@@ -22,10 +22,112 @@
 import cgi
 
 
+class MyTag(object):
+  """Denotes an HTML tag and its content.
+
+  This interface is a minimal subset of BeautifulSoup that we use on tags.
+  The class is a more efficient implementation that runs in 25% of the time
+  (but still about 2.5x longer than hand-constructing HTML).
+  """
+  # pylint: disable=too-few-public-methods
+
+  def __init__(self, name, **kwargs):
+    """Construct a tag
+
+    Args:
+      name: [string] The tag name
+      kwargs: [kwargs] Optional tag key/value attributes
+    """
+    self.__name = name
+    self.__attrs = dict(kwargs)
+    self.__parts = []
+    self.string = ''
+
+  def __setitem__(self, key, value):
+    """Add additional key/value attributes."""
+    self.__attrs[key] = value
+
+  def append(self, part):
+    """Add additional contents (children) to a tag.
+
+    Args:
+      part: [MyTag/String/DocFragment] The thing to add.
+    """
+    self.__parts.append(part)
+
+  def __str__(self):
+    attributes = [' {key}="{value}"'.format(key=key, value=value)
+                  for key, value in self.__attrs.items()]
+    return '<{name}{attributes}>{text}{parts}</{name}>'.format(
```

Comment:
That's what the leading space in the string is for. Doing a ' '.join would only put a space between multiple attributes, not before the initial one. The way I have it avoids special casing and gives you exactly the spaces you need.